[Jason Christopher] iP

jasonchristopher21 opened the pull request:
# Shiba 🦮 

> "So task, much wow" - _Shiba Doge_

Shiba is a task management software dedicated to help you organise your todos, deadlines and events! It's
- Straightforward
- Minimalistic
- ~Easy~ _VERY EASY_ to use

To use **Shiba**, you need to:
1. Download `Duke.jar` file from [here](https://github.com/jasonchristopher21/nus-cs2103t-ip/releases/tag/A-Jar)
2. Run `java -jar a Duke.jar` in your terminal
3. Add your tasks!

Features: 
- [x] Add todo, deadline and event
- [ ] Mark your tasks as done
- [ ] Remove your tasks when needed

If you're a Java Programmer, you may want to modify the Shiba code! Here's the main method:
```java
public static void main(String[] args) {
    Duke duke = new Duke();
    duke.run();
}
```

Commit:
Add Gradle support

jasonchristopher21 committed:
add greeting, echo and bye functionality

jasonchristopher21 committed:
Add indentation and line

jasonchristopher21 committed:
Add add and list

jasonchristopher21 committed:
Add Task class

jasonchristopher21 committed:
Add markAsDone and markAsNotDone functionality

jasonchristopher21 committed:
Add TaskList class

jasonchristopher21 committed:
Add add, mark and unmark

jasonchristopher21 committed:
Edit format and spelling

jasonchristopher21 committed:
Add Todo class

jasonchristopher21 committed:
Add Deadline class

jasonchristopher21 committed:
Apply override on toString methods

jasonchristopher21 committed:
Add Event class

jasonchristopher21 committed:
Integrate todo, event and deadline to Duke and Tasklist

jasonchristopher21 committed:
Add DukeException class

jasonchristopher21 committed:
Add DukeUnspecifiedNumberException class

jasonchristopher21 committed:
Add DukeOutOfRangeException class

jasonchristopher21 committed:
Add DukeInvalidNumberFormatException class

jasonchristopher21 committed:
Add DukeInvalidFormatException class

jasonchristopher21 committed:
Add DukeEmptyDescriptionException class

jasonchristopher21 committed:
Integrate exceptions with Duke and TaskList

jasonchristopher21 committed:
Add delete method

jasonchristopher21 committed:
Integrate DukeUnspecifiedNumberException for invalid mark/unmark/delete inputs

jasonchristopher21 committed:
Implement automated text UI testing

jasonchristopher21 committed:
Add save functionality by creating a Storage class

jasonchristopher21 committed:
Add Date class

jasonchristopher21 committed:
Add date and time functionality

jasonchristopher21 committed:
Implement date parse exception handling

jasonchristopher21 committed:
Fix date toString format

jasonchristopher21 committed:
Merge branch 'branch-Level-8'

LJXSean commented on src/main/java/Date.java:
```diff
@@ -0,0 +1,67 @@
+    protected int time;
+
+    public Date(String str) throws DukeInvalidDateException {
+        String[] splitted = str.split("\\s+");
```
Perhaps you could use a plural form to distinguish a multi-valued variable?

jasonchristopher21 replied:
Right, I'll probably consider renaming it to `splittedWords` instead. Thank you for your comment.

LJXSean commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+                } else if (command.equals("list")) {
+                    printList();
+                } else {
+                    if (command.startsWith("mark")) {
+                        String[] splitted = command.split("\\s+");
+                        Integer index;
+                        if (splitted.length == 1) {
+                            throw new DukeUnspecifiedNumberException("mark");
+                        }
+                        try {
+                            index = Integer.valueOf(splitted[1]);
+                            tasks.markAsDone(index);
+                        } catch (NumberFormatException err) {
+                            throw new DukeInvalidNumberFormatException("mark");
+                        }
```
Is it possible to avoid deep nesting with more than 3 levels of indentation?

jasonchristopher21 replied:
Yup, I just caught up on the SLAP method and the arrowhead style, I would consider implementing it in the next few iterations. Thank you for your comment.

LJXSean commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+        }
+    }
+
+    private static void greetingMessage() {
```
Would using a verb for the method name would be more suitable?

jasonchristopher21 replied:
Right, I'll take it into consideration. Thank you for your comment.

LJXSean commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+        drawLine();
+    }
+
+    public static void indentMessage(String msg) {
```
Using a verb here might be better as well!

jasonchristopher21 replied:
Right, I'll take it into consideration. Thank you for your comment.

LJXSean commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,154 @@
+        f = new File(path);
+    }
+
+    public ArrayList<Task> readFile() {
+        ArrayList<Task> tasks = new ArrayList<>();
+        try {
+            Scanner sc = new Scanner(f);
+            while (sc.hasNext()) {
+                String sentence = sc.nextLine();
+                String[] splitted = sentence.split("\\|");
+                Task t;
+                if (splitted[1] != null) {
+                    if (splitted[1].startsWith("todo")) {
+                        t = addTodo(splitted[1]);
+                    } else if (splitted[1].startsWith("deadline")) {
+                        t = addDeadline(splitted[1]);
+                    } else if (splitted[1].startsWith("event")) {
+                        t = addEvent(splitted[1]);
+                    } else {
+                        t = new Task("dummy");
+                    }
```
Could you avoid using arrowhead nesting if possible?

jasonchristopher21 replied:
Sure thing, thank you very much for your comments, really appreciate them. Will certainly take your comments into consideration for the next iterations! :smile:

LJXSean reviewed (commented):
LGTM, with some nits to fix! 👍

johnrhimawan commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,154 @@
+class Storage {
+
+    private File f;
+    private static final String path = "src/main/data/userTasks.txt";
```
Should this be in uppercase instead since you are declaring a constant?

```suggestion
    private static final String PATH = "src/main/data/userTasks.txt";
```

jasonchristopher21 replied:
Right, I guess I missed this out. Thank you for your comment!

johnrhimawan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+        drawLine();
+    }
+
+    private static void bye() {
```
Perhaps the method name could use a verb like ```sayGoodbye()```? I noticed the same issue in other places as well.

jasonchristopher21 replied:
Right, thank you for your comment, will implement the change in the next iteration.

johnrhimawan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+        drawLine();
+    }
+
+    private static void echo(String msg) {
```
Perhaps a more intuitive method name could be used here?

jasonchristopher21 replied:
Right, thank you for your comment, will implement the change in the next iteration.

johnrhimawan commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,154 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.BufferedWriter;
```
Should ```java.io.BufferedWriter``` be moved to before ```import java.io.File``` to ensure consistent ordering?

jasonchristopher21 replied:
Right, I guess I missed this out when sorting it alphabetically. Thank you for your comment!

johnrhimawan commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+public class Deadline extends Task {
```
Would it be beneficial to organize your code into packages? I noticed the same issue in other files as well.

jasonchristopher21 replied:
I agree with that. I have implemented them under the `A-Packages` tag in the current iteration (as of 9th September), since I had not implemented it last week. Thank you for your comment!

johnrhimawan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+
+    private static boolean isClosed = false;
+    private static TaskList tasks = new TaskList();
+    private static int counter = 0;
```
Perhaps this variable could be removed since it is not used anywhere else?

jasonchristopher21 replied:
Cool, I'll consider removing it for the next iterations.  
Thank you for your comments, I really appreciate them. I'll take them into consideration and implement the changes in the next iteration :smile:

johnrhimawan reviewed (commented):
Overall, your code was easy to read for the most part. I noticed that you followed the coding standard closely in your work. However, there were also several instances where you did not use a verb to name a method and you did not organize your code into packages. Otherwise, LGTM and just a few nits to fix. Excellent work!

jasonchristopher21 committed:
Add and modify classes to adhere with OOP

jasonchristopher21 committed:
Group classes into packages

Group classes from A-MoreOOP into their respective packages and rearrange the order of the imported classes alphabetically

jasonchristopher21 committed:
Change chatbot name and customise messages

jasonchristopher21 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'
Merge is necessary to implement gradle wrapper files and build.gradle to
the app

jasonchristopher21 committed:
Apply Gradle functionality

jasonchristopher21 committed:
Add TodoTest test case

jasonchristopher21 committed:
Add DeadlineTest and DukeExceptionTest

jasonchristopher21 committed:
Fix excess whitespace in Todo argument

jasonchristopher21 committed:
Edit build.grader for building JAR files

jasonchristopher21 committed:
Add Javadoc headers to the public methods

jasonchristopher21 committed:
Merge branch 'branch-A-JavaDoc'

jasonchristopher21 committed:
Fix to comply with coding standard

jasonchristopher21 committed:
Add find functionality

jasonchristopher21 committed:
Merge branch 'branch-Level-9'

jasonchristopher21 committed:
Implement GUI for Duke

jasonchristopher21 committed:
Merge branch 'branch-Level-10'

jasonchristopher21 committed:
Add checkstyle configuration

jasonchristopher21 committed:
Merge branch 'branch-A-CheckStyle'

jasonchristopher21 committed:
Implement Java Assertions

Several methods such as Storage::writeToFile, Date::timeToString,
etc. assumes certain values for it to execute properly.

If it doesn't satisfy these values, then Java might throw an
Unchecked Exception which might mess with the program flow after
GUI is implemented.

Implementing assertions in these points prevents us from messing
with invalid data during the execution of the program, and allows
for better readability on understanding what values are expected
for each parameter.

jasonchristopher21 committed:
Merge pull request #2 from jasonchristopher21/branch-A-Assertions

Implement Java Assertions

jasonchristopher21 committed:
Implement varargs for concatenate

jasonchristopher21 committed:
Fix code quality: implement SLAP and remove arrowhead structure

Code quality is used to enhance readability of the code, and as
taught by the module in the week 5 topics, numerous methods can be
used to fix code quality, including SLAP and arrowhead structure.

Lengthy lines of code evident in `Parser.java` file has been reduced
by implementing a single level of abstraction per method.
Furthermore, codes with arrowhead structure as evident in
TaskList.java has also been fixed by introducing more readable
methods.

Changes to the JavaDoc header has also been made for further
readability, in accordance to the coding style of the module.

jasonchristopher21 committed:
Add gradle.yml file for CI

jasonchristopher21 committed:
Enhance mark, unmark and delete with mass operations

The commands for mark, unmark and delete only supports operating
on one task at a time. However, upon usage, the user may need to
work on multiple tasks at once, for instance marking all the tasks
as done, or deleting a few tasks as done.

Mark, unmark and delete now supports mass operations, as the
keywords "all" or a range of values (e.g. 1-3) can be passed in
as the arguments for the commands.

jasonchristopher21 committed:
Merge pull request #4 from jasonchristopher21/branch-A-CodeQuality

Fix code quality: implement SLAP and remove arrowhead structure

jasonchristopher21 committed:
Merge branch 'master' of https://github.com/jasonchristopher21/ip

jasonchristopher21 committed:
Merge branch 'master' into branch-C-MassOps

jasonchristopher21 committed:
Merge pull request #3 from jasonchristopher21/branch-C-MassOps

Enhance mark, unmark and delete with mass operations

jasonchristopher21 committed:
Merge branch 'master' of https://github.com/jasonchristopher21/ip

jasonchristopher21 committed:
Fix bugs resulting from C-MassOps

jasonchristopher21 committed:
Implement Better GUI

jasonchristopher21 committed:
Implement better personality

Personalise the main window title, and modified several
personalisation comments to reflect the Shiba character

jasonchristopher21 committed:
Add Ui.png file

jasonchristopher21 committed:
Add user guide

jasonchristopher21 committed:
Rename Duke file to Shiba

jasonchristopher21 committed:
Update README.md

jasonchristopher21 committed:
Finalise User Guide

jasonchristopher21 committed:
Update README.md

jasonchristopher21 committed:
Fix CheckStyle issues

jasonchristopher21 committed:
Fix bugs regarding application launch
